

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>velocyto.indexes &mdash; velocyto 0.17.16 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> velocyto
          

          
          </a>

          
            
            
              <div class="version">
                0.17
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome to velocyto.py!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Installation Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../fullapi/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/index.html">Notebooks</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../citing/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">velocyto</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>velocyto.indexes</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for velocyto.indexes</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">typing</span> <span class="k">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">velocyto</span> <span class="k">as</span> <span class="nn">vcy</span>


<div class="viewcode-block" id="TransciptsIndex"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.indexes.TransciptsIndex">[docs]</a><span class="k">class</span> <span class="nc">TransciptsIndex</span><span class="p">:</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;transcipt_models&quot;</span><span class="p">,</span> <span class="s2">&quot;tidx&quot;</span><span class="p">,</span> <span class="s2">&quot;maxtidx&quot;</span><span class="p">]</span>
    <span class="sd">&quot;&quot;&quot;Search help class used to find the transcipt models that a read is spanning/contained into&quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">trascript_models</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">transcipt_models</span> <span class="o">=</span> <span class="n">trascript_models</span>
        <span class="c1"># self.transcipt_models.sort()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tidx</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># index of the current interval</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maxtidx</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">trascript_models</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>

    <span class="o">@</span> <span class="nb">property</span>
    <span class="k">def</span> <span class="nf">scan_not_terminated</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Return false when all the chromosome has been scanned&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">tidx</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">maxtidx</span>
        
<div class="viewcode-block" id="TransciptsIndex.find_overlapping_trascript_models"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.indexes.TransciptsIndex.find_overlapping_trascript_models">[docs]</a>    <span class="k">def</span> <span class="nf">find_overlapping_trascript_models</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">read</span><span class="p">:</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Read</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Set</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Finds all the Transcript models the Read overlaps with</span>

<span class="sd">        Args</span>
<span class="sd">        ----</span>
<span class="sd">        read: vcy.Read</span>
<span class="sd">            the read object to be analyzed</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        matched_transcripts: set of `vcy.TranscriptModel`</span>
<span class="sd">            TranscriptModel the read is overlapping with and values the kind of overlapping</span>
<span class="sd">            it is one of vcy.MATCH_INSIDE (1), vcy.MATCH_OVER5END (2), vcy.MATCH_OVER3END (4)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">matched_transcripts</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">transcipt_models</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;TransciptsIndex </span><span class="si">%s</span><span class="s2"> contains no intervals&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">matched_transcripts</span>

        <span class="n">tmodel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcipt_models</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">tidx</span><span class="p">]</span>  <span class="c1"># current transcript model</span>
        <span class="c1"># Move forward until we find the position we will never search left of again (because the reads are ordered)</span>
        <span class="k">while</span> <span class="bp">self</span><span class="o">.</span><span class="n">scan_not_terminated</span> <span class="ow">and</span> <span class="n">tmodel</span><span class="o">.</span><span class="n">ends_upstream_of</span><span class="p">(</span><span class="n">read</span><span class="p">):</span>
            <span class="c1"># move to the next interval</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tidx</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">tmodel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcipt_models</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">tidx</span><span class="p">]</span>
        
        <span class="c1"># Loop trough the mapping segments of a read (e.g. just one of an internal exon, generally 2 for a splice or intron)</span>
        <span class="k">for</span> <span class="n">segment</span> <span class="ow">in</span> <span class="n">read</span><span class="o">.</span><span class="n">segments</span><span class="p">:</span>
            <span class="c1"># Local search for each segment move a little forward (this just moves a coupple of intervals)</span>
            <span class="n">i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tidx</span>
            <span class="k">while</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">maxtidx</span> <span class="ow">and</span> <span class="n">tmodel</span><span class="o">.</span><span class="n">starts_upstream_of</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                <span class="n">matchtype</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># No match</span>
                <span class="k">if</span> <span class="n">tmodel</span><span class="o">.</span><span class="n">intersects</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                    <span class="c1"># NOTE: do we need to append all the model or the id would be enough?</span>
                    <span class="n">matched_transcripts</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">tmodel</span><span class="p">)</span>
                <span class="c1"># move to the next transcript model</span>
                <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">tmodel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcipt_models</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">matched_transcripts</span></div></div>


<div class="viewcode-block" id="FeatureIndex"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.indexes.FeatureIndex">[docs]</a><span class="k">class</span> <span class="nc">FeatureIndex</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot; Search help class used to find the intervals that a read is spanning &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ivls</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span><span class="p">]</span><span class="o">=</span><span class="p">[])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span> <span class="o">=</span> <span class="n">ivls</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>  <span class="c1"># NOTE: maybe I am sorting twice check what I do upon creation</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># index of the current interval</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maxiidx</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ivls</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="c1"># NOTE needs to be changed to consider situations of identical intervals but different objects</span>

    <span class="o">@</span> <span class="nb">property</span>
    <span class="k">def</span> <span class="nf">last_interval_not_reached</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">maxiidx</span>

<div class="viewcode-block" id="FeatureIndex.reset"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.indexes.FeatureIndex.reset">[docs]</a>    <span class="k">def</span> <span class="nf">reset</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;It set the current feature to the first feature</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span> <span class="o">=</span> <span class="mi">0</span></div>

<div class="viewcode-block" id="FeatureIndex.has_ivls_enclosing"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.indexes.FeatureIndex.has_ivls_enclosing">[docs]</a>    <span class="k">def</span> <span class="nf">has_ivls_enclosing</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">read</span><span class="p">:</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Read</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Finds out if there are intervals that are fully containing all the read segments</span>

<span class="sd">        Args</span>
<span class="sd">        ----</span>
<span class="sd">        read: vcy.Read</span>
<span class="sd">            the read object to be analyzed</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        respones: bool</span>
<span class="sd">            if one has been found</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="n">ivl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iidx</span><span class="p">]</span>  <span class="c1"># current interval</span>
        
        <span class="c1"># Move forward until we find the position we will never search left of again (because the reads are ordered)</span>
        <span class="k">while</span> <span class="bp">self</span><span class="o">.</span><span class="n">last_interval_not_reached</span> <span class="ow">and</span> <span class="n">ivl</span><span class="o">.</span><span class="n">ends_upstream_of</span><span class="p">(</span><span class="n">read</span><span class="p">):</span>
            <span class="c1"># move to the next interval</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">ivl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iidx</span><span class="p">]</span>
        
        <span class="k">for</span> <span class="n">segment</span> <span class="ow">in</span> <span class="n">read</span><span class="o">.</span><span class="n">segments</span><span class="p">:</span>
            <span class="n">segment_matchtype</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># Local search for each segment move a little forward (this just moves a coupple of intervals)</span>
            <span class="n">i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span>
            <span class="n">ivl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iidx</span><span class="p">]</span>
            <span class="k">while</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">maxiidx</span> <span class="ow">and</span> <span class="n">ivl</span><span class="o">.</span><span class="n">doesnt_start_after</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                <span class="n">matchtype</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># No match</span>
                <span class="k">if</span> <span class="n">ivl</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                    <span class="n">matchtype</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">MATCH_INSIDE</span>
                <span class="k">if</span> <span class="n">ivl</span><span class="o">.</span><span class="n">start_overlaps_with_part_of</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>  <span class="c1"># NOTE: should this be elif or it makes sense to allow both?</span>
                    <span class="n">matchtype</span> <span class="o">|=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">MATCH_OVER5END</span>
                <span class="k">if</span> <span class="n">ivl</span><span class="o">.</span><span class="n">end_overlaps_with_part_of</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>  <span class="c1"># NOTE: should this be elif or it makes sense to allow both?</span>
                    <span class="n">matchtype</span> <span class="o">|=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">MATCH_OVER3END</span>

                <span class="n">segment_matchtype</span> <span class="o">|=</span> <span class="n">matchtype</span>
                <span class="c1"># move to the next interval</span>
                <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">ivl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            
            <span class="c1"># If one of the segments does not match inside a repeat we return false</span>
            <span class="k">if</span> <span class="n">segment_matchtype</span> <span class="o">^</span> <span class="n">vcy</span><span class="o">.</span><span class="n">MATCH_INSIDE</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
        <span class="c1"># If I arrive at this point of the code all the segments matched inside</span>
        <span class="k">return</span> <span class="kc">True</span></div>
        
<div class="viewcode-block" id="FeatureIndex.mark_overlapping_ivls"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.indexes.FeatureIndex.mark_overlapping_ivls">[docs]</a>    <span class="k">def</span> <span class="nf">mark_overlapping_ivls</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">read</span><span class="p">:</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Read</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Finds the overlap between Read and Features and mark intronic features if spanned</span>

<span class="sd">        Args</span>
<span class="sd">        ----</span>
<span class="sd">        read: vcy.Read</span>
<span class="sd">            the read object to be analyzed</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Nothing, it marks the vcy.Feature object (is_validated = True) if there is evidence of exon-intron spanning</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># ## TEMPORARY ######</span>
        <span class="c1"># dump_list = []</span>
        <span class="c1">####################</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span>
        <span class="n">feature</span><span class="p">:</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iidx</span><span class="p">]</span>  <span class="c1"># current interval</span>
        <span class="c1"># Move forward until we find the position we will never search left of again (because the reads are ordered)</span>
        <span class="k">while</span> <span class="bp">self</span><span class="o">.</span><span class="n">last_interval_not_reached</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">ends_upstream_of</span><span class="p">(</span><span class="n">read</span><span class="p">):</span>
            <span class="c1"># move to the next interval</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iidx</span><span class="p">]</span>
        
        <span class="c1"># Loop trough the mapping segments of a read (e.g. just one of an internal exon, generally 2 for a splice. for intron???)</span>
        <span class="k">for</span> <span class="n">n_seg</span><span class="p">,</span> <span class="n">segment</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">segments</span><span class="p">):</span>
            <span class="c1"># Local search for each segment move a little forward (this just moves a couple of intervals)</span>
            <span class="n">i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span>
            <span class="n">feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iidx</span><span class="p">]</span>
            <span class="k">while</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">maxiidx</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">doesnt_start_after</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                <span class="c1"># NOTE in this way the checks will be repeated for every clone of an interval in each transcript model</span>
                <span class="c1"># I might want to cache the check results of the previous feature</span>
                <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">kind</span> <span class="o">==</span> <span class="nb">ord</span><span class="p">(</span><span class="s2">&quot;i&quot;</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">end_overlaps_with_part_of</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                        <span class="n">downstream_exon</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">get_downstream_exon</span><span class="p">()</span>
                        <span class="k">if</span> <span class="n">downstream_exon</span><span class="o">.</span><span class="n">start_overlaps_with_part_of</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">is_validated</span> <span class="o">=</span> <span class="kc">True</span>
                            <span class="c1"># ## TEMPORARY ######</span>
                            <span class="c1"># dump_list.append((read, segment, feature, &quot;r&quot;))</span>
                            <span class="c1">####################</span>
                    <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">start_overlaps_with_part_of</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                        <span class="n">upstream_exon</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">get_upstream_exon</span><span class="p">()</span>
                        <span class="k">if</span> <span class="n">upstream_exon</span><span class="o">.</span><span class="n">end_overlaps_with_part_of</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">is_validated</span> <span class="o">=</span> <span class="kc">True</span>
                            <span class="c1"># ## TEMPORARY ######</span>
                            <span class="c1"># dump_list.append((read, segment, feature, &quot;l&quot;))</span>
                            <span class="c1">####################</span>
                    <span class="c1"># if feature.contains(segment):</span>
                    <span class="c1">#     pass  # here the intron is not validated !</span>
                <span class="k">elif</span> <span class="n">feature</span><span class="o">.</span><span class="n">kind</span> <span class="o">==</span> <span class="nb">ord</span><span class="p">(</span><span class="s2">&quot;e&quot;</span><span class="p">):</span>
                    <span class="k">pass</span>  <span class="c1"># here I can pass if I do the proper checks at the intron level</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Unrecognized type of genomic feature {chr(feature.kind)}&quot;</span><span class="p">)</span>

                <span class="c1">#  move to the next interval</span>
                <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="n">i</span><span class="p">]</span></div>

        <span class="c1"># ## TEMPORARY ######</span>
        <span class="c1"># return dump_list</span>
        <span class="c1">####################</span>
        
<div class="viewcode-block" id="FeatureIndex.find_overlapping_ivls"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.indexes.FeatureIndex.find_overlapping_ivls">[docs]</a>    <span class="k">def</span> <span class="nf">find_overlapping_ivls</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">read</span><span class="p">:</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Read</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">SegmentMatch</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot;Finds the possible overlaps between Read and Features and return a 1 read derived mapping record</span>

<span class="sd">        Arguments</span>
<span class="sd">        ---------</span>
<span class="sd">        read: vcy.Read</span>
<span class="sd">            the read object to be analyzed</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        mapping_record: Dict[vcy.TranscriptModel, List[vcy.SegmentMatch]]</span>
<span class="sd">            A record of the mappings by transcript model.</span>
<span class="sd">            Every entry contains a list of segment matches that in turn contains information on the segment and the feature</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>
<span class="sd">        - It is possible that a segment overalps at the same time an exon and an intron (spanning segment)</span>
<span class="sd">        - It is not possible that a segment overalps at the same time two exons. In that case the read is splitted</span>
<span class="sd">        into two segments and  the Read attribute `is_spliced == True`.</span>
<span class="sd">        - Notice that the name of the function might be confousing. if there is a non valid overallapping an empty mappign record will be return</span>
<span class="sd">        - Also notice that returning an empty mapping record will cause the suppression of the counting of the molecule</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">mapping_record</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">SegmentMatch</span><span class="p">]]</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">mapping_record</span>

        <span class="n">feature</span><span class="p">:</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iidx</span><span class="p">]</span>  <span class="c1"># current interval</span>
        <span class="c1"># Move forward until we find the position we will never search left of again (because the reads are ordered)</span>
        <span class="k">while</span> <span class="bp">self</span><span class="o">.</span><span class="n">last_interval_not_reached</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">ends_upstream_of</span><span class="p">(</span><span class="n">read</span><span class="p">):</span>
            <span class="c1"># move to the next interval</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iidx</span><span class="p">]</span>
        
        <span class="c1"># Loop trough the mapping segments of a read (e.g. just one of an internal exon, generally 2 for a splice. for intron???)</span>
        <span class="k">for</span> <span class="n">seg_n</span><span class="p">,</span> <span class="n">segment</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">segments</span><span class="p">):</span>
            <span class="c1"># Local search for each segment move a little forward (this just moves a couple of intervals)</span>
            <span class="n">i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">iidx</span>
            <span class="n">feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">while</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">maxiidx</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">doesnt_start_after</span><span class="p">(</span><span class="n">segment</span><span class="p">):</span>
                <span class="c1"># NOTE in this way the checks will be repeated for every clone of an interval in each transcript model</span>
                <span class="c1"># I might want to cache the check results of the previous feature</span>
                <span class="c1"># it was  if feature.contains(segment) or feature.start_overlaps_with_part_of(segment) or feature.end_overlaps_with_part_of(segment)</span>
                <span class="c1"># but I changed to the more simple</span>
                <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">intersects</span><span class="p">(</span><span class="n">segment</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">&gt;</span> <span class="n">vcy</span><span class="o">.</span><span class="n">MIN_FLANK</span><span class="p">:</span>
                    <span class="n">mapping_record</span><span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">transcript_model</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vcy</span><span class="o">.</span><span class="n">SegmentMatch</span><span class="p">(</span><span class="n">segment</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">is_spliced</span><span class="p">))</span>
                <span class="c1">#  move to the next interval</span>
                <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

        <span class="c1"># NOTE: Removing first the one with less match and then requiring a splicing matching the transcript model is very stringent</span>
        <span class="c1"># It could be that for short ~10bp SKIP sequences the alligner has made a mistake and this might kill the whole molecule</span>
        <span class="c1"># NOTE: the code below is not very efficient</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mapping_record</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Remove transcript models that are suboptimal match</span>
            <span class="c1"># in alternative one could use len(read.segment)</span>
            <span class="n">max_n_segments</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">mapping_record</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="nb">len</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">tm</span><span class="p">,</span> <span class="n">segmatch_list</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">mapping_record</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">segmatch_list</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">max_n_segments</span><span class="p">:</span>
                    <span class="k">del</span> <span class="n">mapping_record</span><span class="p">[</span><span class="n">tm</span><span class="p">]</span>

        <span class="c1"># NOTE: the code below is not very efficient, would be nice to avoid for loops</span>
        <span class="c1"># NOTE: potentailly bad effects: it could kill a lot of molecules if transcript models are not annotated correctly or missing</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mapping_record</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># A SKIP mapping needs to be explainaible by some kind of exon-exon exon-intron junction!</span>
            <span class="c1"># So if it falls internally, the TM needs to be removed from the mapping record</span>
            <span class="k">for</span> <span class="n">tm</span><span class="p">,</span> <span class="n">segmatch_list</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">mapping_record</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
                <span class="k">for</span> <span class="n">sm</span> <span class="ow">in</span> <span class="n">segmatch_list</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">sm</span><span class="o">.</span><span class="n">skip_makes_sense</span><span class="p">:</span>
                        <span class="k">del</span> <span class="n">mapping_record</span><span class="p">[</span><span class="n">tm</span><span class="p">]</span>
                        <span class="k">break</span>

        <span class="k">return</span> <span class="n">mapping_record</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Linnarsson Lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'../../',
              VERSION:'0.17.16',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    

  

  <script type="text/javascript" src="../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>